File: Problems/hacker_rank/Arrays_and_Sorting/20150105_CountingSort2/solution.c
#include <stdio.h>
#include <stdlib.h>

void counting_sort(int* ar, int n)  {
  int* counts = (int*) malloc(sizeof(int) * n);
  int i = 0;
  for ( i = 0; i < n; ++i ) {
    ++*(counts + *(ar + i));
  }
  int idx = 0;
  int* res = (int*) malloc(sizeof(int) * n);
  for ( i = 0; i < n; ++i ) {
    int c = *(counts + i);
    while ( c-- )  {
      *(res + idx++) = i;
    }
  }
  printf("%d", *res);
  for ( i = 1; i < n; ++i ) {
    printf(" %d", *(res + i));
  }
  printf("\n");
  free(res);
}

int main()  {
  int n = 0;
  scanf("%d", &n);
  int i = 0;
  int* ar = (int*) malloc(sizeof(int) * n);
  for ( i = 0; i < n; ++i ) {
    scanf("%d", ar + i);
  }
  counting_sort(ar, n);
  free(ar);
  return 0;
}

File: Problems/hacker_rank/Arrays_and_Sorting/20150105_CountingSort2/Solution.java
import java.util.Scanner;

public class Solution {
  public static int[] countingSort(final int[] ar)  {
    int[] counts = new int[ar.length];
    for ( int i = 0; i < counts.length; ++i )  {
      ++counts[ar[i]];
    }
    int idx = 0;
    int[] res = new int[ar.length];
    for ( int i = 0; i < counts.length; ++i )  {
      int c = counts[i];
      while ( 0 < c-- )  {
        res[idx++] = i;
      }
    }
    return res;
  }

  public static void main(final String[] args)  {
    Scanner sc = new Scanner(System.in);
    final int n = Integer.parseInt(sc.nextLine());
    String[] inps = sc.nextLine().split(" ");
    int[] ar = new int[inps.length];
    for ( int i = 0; i < ar.length; ++i )
      ar[i] = Integer.parseInt(inps[i]);
    int[] res = countingSort(ar);
    System.out.print(res[0]);
    for ( int i = 1; i < res.length; ++i )  {
      System.out.print(" " + res[i]);
    }
    System.out.println();
  }
}
